mush.1
.\" Generated with Ronnjs 0.3.8
.\" http://github.com/kapouer/ronnjs/
.
.TH "MUSH" "1" "December 2022" "" ""
.
.SH "NAME"
\fBmush\fR \-\- Mustache templates for bash
.
.SH "SYNOPSIS"
\fBmush\fR [\-ehV] [\-f <file>] [\-o <file>]
.
.SH "OPTIONS"
  \fB\-f, \-\-file <file>\fR       file to parse
  \fB\-o, \-\-out <file>\fR        output file
  \fB\-e, \-\-escape\fR            escapes html html entities
  \fB\-h, \-\-help\fR              display this message
  \fB\-V, \-\-version\fR           output version
.
.SH "EXAMPLES"
  $ cat file\.ms | FOO=BAR mush
  $ VALUE=123 mush \-f file\.ms \-o file
  $ echo "Today\'s date is {{DATE}}" | DATE=`date +%D` mush
  $ cat \./template\.ms | VAR=VALUE mush
.
.SH "USAGE"
  Suppose you have a template file:
.
.P
  \./template\.ms
.
.P
  \fB
  VAR={{VAR}}
  \fR
.
.P
  You can compile it like such: 
.
.P
  \fB
  $ cat \./template\.ms | VAR=VALUE mush
  VAR=123
  \fR
.
.P
  You can utilize stdin in the same way with \fBecho\fR
.
.P
  \fB
  $ echo "Today\'s date is {{DATE}}" | DATE=`date +%D` mush
  Today\'s date is 12/17/13
  \fR
.
.P
  Variables are passed to the view vie environment variable
  definition\. Due to the way variables are scoped to the
  templates\. All environment variables are available to the
  template\. This includes variables like \fB$HOME\fR and \fB$USER\fR
.
.P
  Compile a HTML file with partial:
.
.P
 index\.html\.ms (layout)
.
.P
  \fB
  <!doctype html>
  <html>
    <head>
      <meta http\-equiv="content\-type" content="text/html; charset=utf\-8"/>
      <title>{{title}}</title>
      <link rel="stylesheet" href="{{main_css}}" type="text/css" />
      <script type="text/javascript" src="{{main_js}}" charset="utf\-8"></script>
    </head>
    <body>
      <div id="app">
        {{content}}
      </div>
    </body>
  </html>
  \fR
.
.P
  page\.html\.ms (partial)
.
.P
  \fB
  <div id="{{name}}" class="page"> This is the {{name}} page </div>
  \fR
.
.P
  \fB
  $ cat index\.html\.ms | \\
    title="Awesome Web Site" \\
    main_css="/css/main\.css" \\
    main_js="/js/main\.js" \\
    content="`cat page\.html\.ms | name="home" mush`" \\
    mush
  \fR
.
.P
  This will yield:
.
.P
  \fB
  <!doctype html>
  <html>
    <head>
      <meta http\-equiv="content\-type" content="text/html; charset=utf\-8"/>
      <title>foo</title>
      <link rel="stylesheet" href="/css/main\.css" type="text/css" />
      <script type="text/javascript" src="/js/app\.js" charset="utf\-8"></script>
    </head>
    <body>
      <div id="app">
        <div id="home" class="page"> This is the home page </div>
      </div>
    </body>
  </html>
  \fR
.
.SH "AUTHOR"
.
.IP "\(bu" 4
Joseph Werle \fIjoseph\.werle@gmail\.com\fR
.
.IP "" 0
.
.SH "REPORTING BUGS"
.
.IP "\(bu" 4
https://github\.com/jwerle/mush/issues
.
.IP "" 0
.
.SH "SEE ALSO"
.
.IP "\(bu" 4
http://mustache\.github\.io/
.
.IP "\(bu" 4
https://github\.com/jwerle/mush
.
.IP "" 0
.
.SH "LICENSE"
  MIT (C) Copyright Joseph Werle 2013-2022


mush.sh
#!/usr/bin/env bash

mush_version () {
  echo "0.1.0"
}

mush_usage () {
  echo "usage: mush [-ehV] [-f <file>] [-o <file>]"

  if [ "$1" = "1" ]; then
    echo
    echo "examples:"
    echo "  $ cat file.ms | FOO=BAR mush"
    echo "  $ VALUE=123 mush -f file.ms -o file"
    echo "  $ echo \"Today's date is {{DATE}}\" | DATE=\`date +%D\` mush"
    echo "  $ cat ./template.ms | VAR=VALUE mush"
    echo
    echo "options:"
    echo "  -f, --file <file>       file to parse"
    echo "  -o, --out <file>        output file"
    echo "  -e, --escape            escapes html html entities"
    echo "  -h, --help              display this message"
    echo "  -V, --version           output version"
  fi
}

mush () {
  # shellcheck disable=SC2034
  local SELF="$0"
  # shellcheck disable=SC2034
  local NULL=/dev/null
  # shellcheck disable=SC2034
  local STDIN=0
  local STDOUT=1
  local STDERR=2
  local LEFT_DELIM="{{"
  local RIGHT_DELIM="}}"
  # shellcheck disable=SC2034
  local INDENT_LEVEL="  "
  local ESCAPE=0
  local ENV=""
  local out=">&$STDOUT"

  ENV="$(env)"

  ## parse opts
  while true; do
    arg="$1"

    if [ "" = "$1" ]; then
      break;
    fi

    if [ "${arg:0:1}" != "-" ]; then
      shift
      continue
    fi

    case $arg in
      -f|--file)
        file="$2";
        shift 2;
        ;;
      -o|--out)
        out="> $2";
        shift 2;
        ;;
      -e|--escape)
        ESCAPE=1
        shift
        ;;
      -h|--help)
        mush_usage 1
        exit 1
        ;;
      -V|--version)
        mush_version
        exit 0
        ;;
      *)
        {
          echo "unknown option \`$arg'"
        } >&$STDERR
        mush_usage
        exit 1
        ;;
    esac
  done

  ## read each line
  while IFS= read -r line; do
    printf '%q\n' "${line}" | {
        ## read each ENV variable
        echo "$ENV" | {
          while read -r var; do
            ## split each ENV variable by '='
            ## and parse the line replacing
            ## occurrence of the key with
            ## guarded by the values of
            ## `LEFT_DELIM' and `RIGHT_DELIM'
            ## with the value of the variable
            case "$var" in
              (*"="*)
                key=${var%%"="*}
                val=${var#*"="*}
                ;;

              (*)
                key=$var
                val=
                ;;
            esac

            line="${line//${LEFT_DELIM}$key${RIGHT_DELIM}/$val}"
          done

          if [ "1" = "$ESCAPE" ]; then
            line="${line//&/&amp;}"
            line="${line//\"/&quot;}"
            line="${line//\</&lt;}"
            line="${line//\>/&gt;}"
          fi

          ## output to stdout
          echo "$line" | {
            ## parse undefined variables
            sed -e "s#${LEFT_DELIM}[A-Za-z]*${RIGHT_DELIM}##g" | \
            ## parse comments
            sed -e "s#${LEFT_DELIM}\!.*${RIGHT_DELIM}##g" | \
            ## escaping
            sed -e 's/\\\"/""/g'
          };
        }
    };
  done
}


if [[ "${BASH_SOURCE[0]}" != "$0" ]]; then
  export -f mush
else
  if [ ! -t 0 ]; then
    eval "mush $out"
  elif [ -n "$file" ]; then
    eval "cat $file | mush $out"
  elif (( $# > 0 )); then
    mush "$@"
  else
    mush_usage
    exit 1
  fi
  exit $?
fi

